modules/opensles/recorder.c

```diff
@@ -107,6 +112,49 @@ static int createAudioRecorder(struct ausrc_st *st, struct ausrc_prm *prm)
 		return ENODEV;
 	}
 
+	SLAndroidConfigurationItf player_config;
+	r = (*st->recObject)->GetInterface(st->recObject,
+									   SL_IID_ANDROIDCONFIGURATION,
+									   &player_config);
+	if (SL_RESULT_SUCCESS == r) {
+		SLint32 stream_type = SL_ANDROID_RECORDING_PRESET_VOICE_COMMUNICATION;
+
+		(*player_config)->SetConfiguration(player_config,
+										   SL_ANDROID_KEY_RECORDING_PRESET, &stream_type, sizeof(SLint32));
+
+		SLuint32 performance = SL_ANDROID_PERFORMANCE_NONE;
+		(*player_config)->SetConfiguration(player_config,
+										   SL_ANDROID_KEY_PERFORMANCE_MODE, &performance, sizeof(SLuint32));
+	}
+
+
+	SLAndroidAcousticEchoCancellationItf player_aec;
+	r = (*st->recObject)->GetInterface(st->recObject,
+									   SL_IID_ANDROIDACOUSTICECHOCANCELLATION,
+									   &player_aec);
+	if (SL_RESULT_SUCCESS == r) {
+		SLboolean enable = SL_BOOLEAN_TRUE;
+		(*player_aec)->SetEnabled(player_aec, enable);
+	}
+
+	SLAndroidAutomaticGainControlItf player_gain;
+	r = (*st->recObject)->GetInterface(st->recObject,
+									   SL_IID_ANDROIDAUTOMATICGAINCONTROL,
+									   &player_gain);
+	if (SL_RESULT_SUCCESS == r) {
+		SLboolean enable = SL_BOOLEAN_TRUE;
+		(*player_gain)->SetEnabled(player_gain, enable);
+	}
+
+	SLAndroidNoiseSuppressionItf player_noise;
+	r = (*st->recObject)->GetInterface(st->recObject,
+									   SL_IID_ANDROIDAUTOMATICGAINCONTROL,
+									   &player_noise);
+	if (SL_RESULT_SUCCESS == r) {
+		SLboolean enable = SL_BOOLEAN_TRUE;
+		(*player_noise)->SetEnabled(player_noise, enable);
+	}
```

Comment:
AEC, AEG and NS should be configurable, either via baresip core or only for this module.

For example running two instances of AEC (this one and webrtc_aec) can lead
to bad audio and/or interference.

Reply:
Yes, they should be configurable.  For example, here is how AEC is set in Google test code:
```
    /* Enable AEC if requested */
    if (aec) {
        SLAndroidAcousticEchoCancellationItf aecItf;
        result = (*recorder)->GetInterface(
                recorder, SL_IID_ANDROIDACOUSTICECHOCANCELLATION, (void*)&aecItf);
        printf("AEC is %savailable\n", SL_RESULT_SUCCESS == result ? "" : "not ");
        if (SL_RESULT_SUCCESS == result) {
            result = (*aecItf)->SetEnabled(aecItf, true);
            ExitOnError(result);
            SLboolean enabled;
            result = (*aecItf)->IsEnabled(aecItf, &enabled);
            ExitOnError(result);
            printf("AEC is %s\n", enabled ? "enabled" : "not enabled");
        }
    }
```
Same kind if debug messages would be useful also barersip module.

Reply:
I added info message here:
```
	SLAndroidAcousticEchoCancellationItf player_aec;
	r = (*st->recObject)->
		GetInterface(st->recObject,
			     SL_IID_ANDROIDACOUSTICECHOCANCELLATION,
			     &player_aec);
	info("AEC is %savailable\n", SL_RESULT_SUCCESS == r ? "" : "not ");
```
And on my Android 10 device got:
```
10-02 09:59:01.086  5275  5321 D Baresip Lib: opensles: opening recorder 16000Hz, 1channels
10-02 09:59:01.086  5275  5321 D Baresip Lib: AEC is not available
```
So are these features only available on old Android devices?

Reply:
Yes, on my Chinese tablet log - is enabled true